dds/idl/marshal_generator.cpp

```diff
@@ -3017,6 +3546,55 @@ bool marshal_generator::gen_union(AST_Union* node, UTL_ScopedName* name,
 
   const bool not_final = exten != extensibilitykind_final;
 
+  {
+    Function set_default("set_default", "void");
+    set_default.addArg("stru", cxx + "&");
+    set_default.endArgs();
+    be_global->impl_ << "  " << scoped(discriminator->name()) << " temp;\n";
+    be_global->impl_ << type_to_default(discriminator, "  temp");
+    be_global->impl_ << "  stru._d(temp);\n";
+    AST_Type* disc_type = resolveActualType(discriminator);
+    Classification disc_cls = classify(disc_type);
+    ACE_CDR::ULong default_enum_val = 0;
+    if (disc_cls & CL_ENUM) {
+      AST_Enum* enu = dynamic_cast<AST_Enum*>(disc_type);
+      UTL_ScopeActiveIterator i(enu, UTL_Scope::IK_decls);
+      AST_EnumVal *item = dynamic_cast<AST_EnumVal*>(i.item());
+      default_enum_val = item->constant_value()->ev()->u.eval;
+    }
+    bool found = false;
+    for (std::vector<AST_UnionBranch*>::const_iterator itr = branches.begin(); itr < branches.end() && !found; itr++) {
+      AST_UnionBranch* branch = *itr;
+      for (unsigned i = 0; i < branch->label_list_length(); i++) {
+        AST_UnionLabel* ul = branch->label(i);
+        if (ul->label_kind() != AST_UnionLabel::UL_default) {
+          AST_Expression::AST_ExprValue* ev = branch->label(i)->label_val()->ev();
+          if ((ev->et == AST_Expression::EV_enum && ev->u.eval == default_enum_val) ||
+              (ev->et == AST_Expression::EV_short && ev->u.sval == 0) ||
+              (ev->et == AST_Expression::EV_ushort && ev->u.usval == 0) ||
+              (ev->et == AST_Expression::EV_long && ev->u.lval == 0) ||
+              (ev->et == AST_Expression::EV_ulong && ev->u.ulval == 0) ||
+              (ev->et == AST_Expression::EV_longlong && ev->u.llval == 0) ||
+              (ev->et == AST_Expression::EV_ulonglong && ev->u.ullval == 0) ||
+              (ev->et == AST_Expression::EV_float && ev->u.fval == 0) ||
+              (ev->et == AST_Expression::EV_double && ev->u.dval == 0) ||
+              (ev->et == AST_Expression::EV_longdouble && ev->u.sval == 0) ||
+              (ev->et == AST_Expression::EV_char && ev->u.cval == 0) ||
+              (ev->et == AST_Expression::EV_wchar && ev->u.wcval == 0) ||
+              (ev->et == AST_Expression::EV_octet && ev->u.oval == 0) ||
+              (ev->et == AST_Expression::EV_bool && ev->u.bval == 0))
+          {
+            be_global->impl_ << "  " << scoped(branch->field_type()->name()) << " btemp;\n";
+            be_global->impl_ << type_to_default(branch->field_type(), "  btemp");
+            be_global->impl_ << "  stru." << branch->local_name()->get_string() << "(btemp);\n";
```

Comment:
I'm getting a warning from the code generated here in `tests/DCPS/Compiler/C++11/idl_test1_lib/FooDefTypeSupportImpl.cpp`:

``` C++
   2459 void set_default(Xyz::AUnion& stru)                                                                                                                                                                        
   2460 {                                                                                                                                                                                                          
   2461   ACE_UNUSED_ARG(stru);                                                                                                                                                                                    
   2462   Xyz::ColorX temp;                                                                                                                                                                                        
   2463   temp = Xyz::ColorX::redx;                                                                                                                                                                                
   2464   stru._d(temp);                                                                                                                                                                                           
   2465   char * btemp;                                                                                                                                                                                            
⚠  2466   btemp = ""; clang: ISO C++11 does not allow conversion from string literal to 'char *'                                                                                                                   
   2467   stru.rsv(btemp);                                                                                                                                                                                         
   2468 } 
```